#!/bin/bash

module load BCFtools/1.14-GCC-11.2.0

# run exactly here at the script location, even if called from somewhere else
SCRIPTDIR=$(cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
cd $SCRIPTDIR

# location to write prefiltered output to
OUTDIR="variant_calls_prefiltered/sarek_GATK_hg38_plus_Carvykti"
mkdir -p $OUTDIR

# locations of sarek pipeline output, relative to this script
BWA_OUT="pipeline_runs/initial_bwamem2/sarek_GATK_hg38_plus_Carvykti_bwamem2/"
BWA_OUT_ANNO=$BWA_OUT/annotation
BWA_OUT_RAW=$BWA_OUT/variant_calling
DM_OUT="pipeline_runs/initial_dragmap/sarek_GATK_hg38_plus_Carvykti_dragmap/"
DM_OUT_ANNO=$DM_OUT/annotation
DM_OUT_RAW=$DM_OUT/variant_calling

# using bcftools, discard everything that is not annotated as PASS where applicable
# also use bcftools norm to normalize indels and split multiallelic sites into multiple rows,
# so that we have an easier time intersecting variants from multiple callers.

### Apherese Sample S1
echo "Filtering Apherese (Run 1: Sample S1)"

# remove variants below 20x and above 200x coverage
# except for manta, since depth information is not so trivially accessible for large SVs and manta has filters in place

bcftools view -f .,PASS -e 'FORMAT/DP<20 || FORMAT/DP>200' "$BWA_OUT_ANNO/deepvariant/Apherese_CARneg_CD5pos/Apherese_CARneg_CD5pos.deepvariant_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/Apherese_CARneg_CD5pos.bwamem2.deepvariant_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<20 || FORMAT/DP>200' "$DM_OUT_ANNO/deepvariant/Apherese_CARneg_CD5pos/Apherese_CARneg_CD5pos.deepvariant_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/Apherese_CARneg_CD5pos.dragmap.deepvariant_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<20 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/Apherese_CARneg_CD5pos/Apherese_CARneg_CD5pos.strelka.variants_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/Apherese_CARneg_CD5pos.bwamem2.strelka.variants_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<20 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/Apherese_CARneg_CD5pos/Apherese_CARneg_CD5pos.strelka.variants_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/Apherese_CARneg_CD5pos.dragmap.strelka.variants_VEP.filtered.vcf.gz &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/Apherese_CARneg_CD5pos.bwamem2.manta.diploid_sv_VEP.filtered.vcf.gz "$BWA_OUT_ANNO/manta/Apherese_CARneg_CD5pos/Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz" &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/Apherese_CARneg_CD5pos.dragmap.manta.diploid_sv_VEP.filtered.vcf.gz "$DM_OUT_ANNO/manta/Apherese_CARneg_CD5pos/Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz"

### PB_CARpos_CD5dim Sample S2
echo "Filtering CARpos_CD5dim (Run 1: Sample S2)"

# remove variants below 10x and above 200x coverage
# except for manta, since depth information is not so trivially accessible for large SVs and manta has filters in place. For somatic manta output we need to apply PASS filtering.

bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/mutect2/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/mutect2/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.manta.diploid_sv_VEP.filtered.vcf.gz "$BWA_OUT_ANNO/manta/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz" &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.manta.diploid_sv_VEP.filtered.vcf.gz "$DM_OUT_ANNO/manta/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz"

### P2251_CARpos_CD5dim Sample S1
echo "Filtering CARpos_CD5dim (Run 2: Sample S1 (P2251))"

# remove variants below 10x and above 200x coverage
# except for manta, since depth information is not so trivially accessible for large SVs and manta has filters in place. For somatic manta output we need to apply PASS filtering.

bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/mutect2/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/mutect2/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.bwamem2.manta.diploid_sv_VEP.filtered.vcf.gz "$BWA_OUT_ANNO/manta/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz" &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.dragmap.manta.diploid_sv_VEP.filtered.vcf.gz "$DM_OUT_ANNO/manta/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos/P2251_CARpos_CD5dim_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz"

### Both CARpos_CD5dim
echo "Filtering CARpos_CD5dim_merged (Run 1: Sample S2 + Run 2: Sample S1 (P2251))"

# remove variants below 10x and above 200x coverage
# except for manta, since depth information is not so trivially accessible for large SVs and manta has filters in place. For somatic manta output we need to apply PASS filtering.

bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/mutect2/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.bwamem2.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/mutect2/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.dragmap.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.bwamem2.manta.diploid_sv_VEP.filtered.vcf.gz "$BWA_OUT_ANNO/manta/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz" &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.dragmap.manta.diploid_sv_VEP.filtered.vcf.gz "$DM_OUT_ANNO/manta/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5dim_merged_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz"

### Both CARpos_CD5pos
echo "Filtering CARpos_CD5pos_merged (Run 1: Sample S3 + Run 2: Sample S2)"

# remove variants below 10x and above 200x coverage
# except for manta, since depth information is not so trivially accessible for large SVs and manta has filters in place. For somatic manta output we need to apply PASS filtering.

bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/mutect2/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/mutect2/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.manta.diploid_sv_VEP.filtered.vcf.gz "$BWA_OUT_ANNO/manta/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz" &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.manta.diploid_sv_VEP.filtered.vcf.gz "$DM_OUT_ANNO/manta/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARpos_CD5pos_merged_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz"

### Both CARneg_CD5pos
echo "Filtering CARneg_CD5pos_merged (Run 1: Sample S4 + Run 2: Sample S3)"

# remove variants below 10x and above 200x coverage
# except for manta, since depth information is not so trivially accessible for large SVs and manta has filters in place. For somatic manta output we need to apply PASS filtering.

bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/mutect2/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/mutect2/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.mutect2.filtered_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.mutect2_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_snvs_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_snvs_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$BWA_OUT_ANNO/strelka/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view -f .,PASS -e 'FORMAT/DP<10 || FORMAT/DP>200' "$DM_OUT_ANNO/strelka/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.strelka.somatic_indels_VEP.ann.vcf.gz" | bcftools norm --threads 4 --fasta-ref "/mnt/ribolution/user_worktmp/florian.peter.grosse/scratch/igenomes/Homo_sapiens_plus_Carvykti/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38_plus_Carvykti.fasta" -m - -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.strelka.somatic_indels_VEP.filtered.vcf.gz &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.bwamem2.manta.diploid_sv_VEP.filtered.vcf.gz "$BWA_OUT_ANNO/manta/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz" &
bcftools view --threads 4 -f .,PASS -o $OUTDIR/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.dragmap.manta.diploid_sv_VEP.filtered.vcf.gz "$DM_OUT_ANNO/manta/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos/PB_CARneg_CD5pos_merged_vs_Apherese_CARneg_CD5pos.manta.diploid_sv_VEP.ann.vcf.gz"
